### ISSUE TITLE
Integration Test Framework do not work out the box. Not compatible with PhpUnit 9.3.x

### ISSUE THREAD

**Issue by revgovmade** [magento/magento2#30146]
### Preconditions
1. Magento 2.4.0

### Steps to reproduce
1. Run Integration Test Framework on a new test or existing Magento 2 test.
2. cd into the following folder: `dev/tests/integration`
3. `../../../vendor/phpunit/phpunit/phpunit --configuration phpunit.xml`

### Expected Result
1. Tests run.

### Actual Result
```
PHP Fatal error:  Uncaught Error: Class 'PHPUnit\TextUI\Configuration\Registry' not found in /var/www/dev/tests/integration/testsuite/Magento/IntegrationTest.php:32
Stack trace:
#0 [internal function]: Magento\IntegrationTest::suite('Magento\\Integra...')
#1 /var/www/vendor/phpunit/phpunit/src/Framework/TestSuite.php(476): ReflectionMethod->invoke(NULL, 'Magento\\Integra...')
#2 /var/www/vendor/phpunit/phpunit/src/TextUI/XmlConfiguration/TestSuite/TestSuiteMapper.php(65): PHPUnit\Framework\TestSuite->addTestFile('/var/www/dev/te...')
#3 /var/www/vendor/phpunit/phpunit/src/TextUI/Command.php(389): PHPUnit\TextUI\XmlConfiguration\TestSuiteMapper->map(Object(PHPUnit\TextUI\XmlConfiguration\TestSuiteCollection), '')
#4 /var/www/vendor/phpunit/phpunit/src/TextUI/Command.php(116): PHPUnit\TextUI\Command->handleArguments(Array)
#5 /var/www/vendor/phpunit/phpunit/src/TextUI/Command.php(101): PHPUnit\TextUI\Command->run(Array, true)
#6 /var/www/vendor/phpunit/phpunit/phpunit(61): PHPUnit\TextUI\Command::main()
#7 {main}

Next PHPUnit\T in /var/www/vendor/phpunit/phpunit/src/TextUI/Command.php on line 103

Fatal error: Uncaught Error: Class 'PHPUnit\TextUI\Configuration\Registry' not found in /var/www/dev/tests/integration/testsuite/Magento/IntegrationTest.php:32
Stack trace:
#0 [internal function]: Magento\IntegrationTest::suite('Magento\\Integra...')
#1 /var/www/vendor/phpunit/phpunit/src/Framework/TestSuite.php(476): ReflectionMethod->invoke(NULL, 'Magento\\Integra...')
#2 /var/www/vendor/phpunit/phpunit/src/TextUI/XmlConfiguration/TestSuite/TestSuiteMapper.php(65): PHPUnit\Framework\TestSuite->addTestFile('/var/www/dev/te...')
#3 /var/www/vendor/phpunit/phpunit/src/TextUI/Command.php(389): PHPUnit\TextUI\XmlConfiguration\TestSuiteMapper->map(Object(PHPUnit\TextUI\XmlConfiguration\TestSuiteCollection), '')
#4 /var/www/vendor/phpunit/phpunit/src/TextUI/Command.php(116): PHPUnit\TextUI\Command->handleArguments(Array)
#5 /var/www/vendor/phpunit/phpunit/src/TextUI/Command.php(101): PHPUnit\TextUI\Command->run(Array, true)
#6 /var/www/vendor/phpunit/phpunit/phpunit(61): PHPUnit\TextUI\Command::main()
#7 {main}

Next PHPUnit\T in /var/www/vendor/phpunit/phpunit/src/TextUI/Command.php on line 103
PHP Fatal error:  Uncaught Exception: sh: tasklist.exe: not found in /var/www/vendor/magento/framework/Shell.php:63
Stack trace:
#0 /var/www/dev/tests/integration/framework/Magento/TestFramework/Helper/Memory.php(92): Magento\Framework\Shell->execute('tasklist.exe /f...', Array)
#1 /var/www/dev/tests/integration/framework/Magento/TestFramework/Helper/Memory.php(58): Magento\TestFramework\Helper\Memory->_getWinProcessMemoryUsage(47150)
#2 /var/www/dev/tests/integration/framework/Magento/TestFramework/MemoryLimit.php(127): Magento\TestFramework\Helper\Memory->getRealMemoryUsage()
#3 /var/www/dev/tests/integration/framework/Magento/TestFramework/MemoryLimit.php(59): Magento\TestFramework\MemoryLimit->_getUsage()
#4 /var/www/dev/tests/integration/framework/Magento/TestFramework/Bootstrap/Memory.php(50): Magento\TestFramework\MemoryLimit->printStats()
#5 [internal function]: Magento\TestFramework\Bootstrap\Memory->displayStats()
#6 {main}

Next Magento\Framework\Exception\LocalizedException: Command returned non-zero exit code:
`ta in /var/www/vendor/magento/framework/Shell.php on line 64

Fatal error: Uncaught Exception: sh: tasklist.exe: not found in /var/www/vendor/magento/framework/Shell.php:63
Stack trace:
#0 /var/www/dev/tests/integration/framework/Magento/TestFramework/Helper/Memory.php(92): Magento\Framework\Shell->execute('tasklist.exe /f...', Array)
#1 /var/www/dev/tests/integration/framework/Magento/TestFramework/Helper/Memory.php(58): Magento\TestFramework\Helper\Memory->_getWinProcessMemoryUsage(47150)
#2 /var/www/dev/tests/integration/framework/Magento/TestFramework/MemoryLimit.php(127): Magento\TestFramework\Helper\Memory->getRealMemoryUsage()
#3 /var/www/dev/tests/integration/framework/Magento/TestFramework/MemoryLimit.php(59): Magento\TestFramework\MemoryLimit->_getUsage()
#4 /var/www/dev/tests/integration/framework/Magento/TestFramework/Bootstrap/Memory.php(50): Magento\TestFramework\MemoryLimit->printStats()
#5 [internal function]: Magento\TestFramework\Bootstrap\Memory->displayStats()
#6 {main}

Next Magento\Framework\Exception\LocalizedException: Command returned non-zero exit code:
`ta in /var/www/vendor/magento/framework/Shell.php on line 64
```

---
Please provide [Severity](https://devdocs.magento.com/guides/v2.3/contributor-guide/contributing.html#backlog) assessment for the Issue as Reporter. This information will help during Confirmation and Issue triage processes.

- [ ] Severity: **S0** _- Affects critical data or functionality and leaves users without workaround._
- [ ] Severity: **S1** _- Affects critical data or functionality and forces users to employ a workaround._
- [x] Severity: **S2** _- Affects non-critical data or functionality and forces users to employ a workaround._
- [ ] Severity: **S3** _- Affects non-critical data or functionality and does not force users to employ a workaround._
- [ ] Severity: **S4** _- Affects aesthetics, professional look and feel, “quality” or “usability”._

**Comment by ihor-sviziev** [comment]
Hi @revgovmade,
Could you add steps to reproduce?

**Comment by revgovmade** [comment]
Hi @ihor-sviziev, Steps to Reproduce are in the original bug report above. You just have to run the integration tests out the box and they fail. 

It appears the following class is not compatible with PhpUnit 9.x.x: https://github.com/magento/magento2/blob/2.4-develop/dev/tests/integration/testsuite/Magento/IntegrationTest.php

The following three files are referenced, but do not ship with PhpUnit 9.x.x:
- PHPUnit\TextUI\Configuration\Registry;
- PHPUnit\TextUI\Configuration\TestSuiteCollection;
- PHPUnit\TextUI\Configuration\TestSuiteMapper;

**Comment by ihor-sviziev** [comment]
@revgovmade ok. which command have you executed in order to run tests?

**Comment by revgovmade** [comment]
First I cd into the following folder: `dev/tests/integration`
Then run: `../../../vendor/phpunit/phpunit/phpunit --configuration phpunit.xml`

**Comment by ihor-sviziev** [comment]
Just tried to run this test on 2.4-develop, and seems like you're right - it doesn't work.

But I actually got a bit different error:
```
PHP Fatal error:  Uncaught TypeError: Argument 1 passed to Magento\TestFramework\Workaround\Override\Config\RelationsCollector::getRelations() must be of the type string, null given, called in /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Override/Config/RelationsCollector.php on line 53 and defined in /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Override/Config/RelationsCollector.php:47
Stack trace:
#0 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Override/Config/RelationsCollector.php(53): Magento\TestFramework\Workaround\Override\Config\RelationsCollector->getRelations(NULL)
#1 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Override/Config/RelationsCollector.php(53): Magento\TestFramework\Workaround\Override\Config\RelationsCollector->getRelations('PHPUnit\\Framewo...')
#2 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Override/Config/RelationsCollector.php(38):  in /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Override/Config/RelationsCollector.php on line 47

Fatal error: Uncaught TypeError: Argument 1 passed to Magento\TestFramework\Workaround\Override\Config\RelationsCollector::getRelations() must be of the type string, null given, called in /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Override/Config/RelationsCollector.php on line 53 and defined in /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Override/Config/RelationsCollector.php:47
Stack trace:
#0 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Override/Config/RelationsCollector.php(53): Magento\TestFramework\Workaround\Override\Config\RelationsCollector->getRelations(NULL)
#1 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Override/Config/RelationsCollector.php(53): Magento\TestFramework\Workaround\Override\Config\RelationsCollector->getRelations('PHPUnit\\Framewo...')
#2 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Override/Config/RelationsCollector.php(38):  in /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Override/Config/RelationsCollector.php on line 47

=== Memory Usage System Stats ===
Memory usage (OS):      88.44M (175.13% of 50.50M reported by PHP, 1.08% of configured 8192.00M limit)
Estimated memory leak:  37.94M (42.90% of used memory)
```
![image](https://user-images.githubusercontent.com/1873745/94000302-acdccb80-fd9f-11ea-8ca1-35ecd99f1fe3.png)


Also on 2.4.0 I was able to reproduce a bit different, but it also fails:
```
www-data@m24develop-php-fpm:/var/www/html/dev/tests/integration$ ../../../vendor/phpunit/phpunit/phpunit --configuration phpunit.xml ./testsuite/Magento/IntegrationTest.php
PHPUnit 9.1.5 by Sebastian Bergmann and contributors.

Warning:       Using a custom test suite loader is deprecated

PHP Fatal error:  Uncaught InvalidArgumentException: Unable to unserialize value. Error: Syntax error in /var/www/html/lib/internal/Magento/Framework/Serialize/Serializer/Json.php:39
Stack trace:
#0 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Cleanup/StaticProperties.php(175): Magento\Framework\Serialize\Serializer\Json->unserialize('a:0:{}')
#1 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Cleanup/StaticProperties.php(231): Magento\TestFramework\Workaround\Cleanup\StaticProperties::backupStaticVariables()
#2 [internal function]: Magento\TestFramework\Workaround\Cleanup\StaticProperties->startTestSuite()
#3 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/EventManager.php(49): call_user_func_array(Array, Array)
#4 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Event/PhpUnit.php(105): Magento\TestFramework\EventManager->fireEvent('startTestSuite')
#5 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestResult.php(384):  in /var/www/html/lib/internal/Magento/Framework/Serialize/Serializer/Json.php on line 39

Fatal error: Uncaught InvalidArgumentException: Unable to unserialize value. Error: Syntax error in /var/www/html/lib/internal/Magento/Framework/Serialize/Serializer/Json.php:39
Stack trace:
#0 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Cleanup/StaticProperties.php(175): Magento\Framework\Serialize\Serializer\Json->unserialize('a:0:{}')
#1 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Workaround/Cleanup/StaticProperties.php(231): Magento\TestFramework\Workaround\Cleanup\StaticProperties::backupStaticVariables()
#2 [internal function]: Magento\TestFramework\Workaround\Cleanup\StaticProperties->startTestSuite()
#3 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/EventManager.php(49): call_user_func_array(Array, Array)
#4 /var/www/html/dev/tests/integration/framework/Magento/TestFramework/Event/PhpUnit.php(105): Magento\TestFramework\EventManager->fireEvent('startTestSuite')
#5 /var/www/html/vendor/phpunit/phpunit/src/Framework/TestResult.php(384):  in /var/www/html/lib/internal/Magento/Framework/Serialize/Serializer/Json.php on line 39

=== Memory Usage System Stats ===
Memory usage (OS):      679.04M (105.60% of 643.00M reported by PHP, 8.29% of configured 8192.00M limit)
Estimated memory leak:  36.04M (5.31% of used memory)
```

**Comment by magento-engcom-team** [comment]
:white_check_mark: Confirmed by @ihor-sviziev 
 Thank you for verifying the issue. Based on the provided information internal tickets `MC-37866` were created

**Issue Available**: @ihor-sviziev, _You will be automatically unassigned. Contributors/Maintainers can claim this issue to continue. To reclaim and continue work, reassign the ticket to yourself._

**Comment by ihor-sviziev** [comment]
@revgovmade thank you for your report, I acknowledged it!

**Comment by ihor-sviziev** [comment]
@revgovmade could you provide which phpunit version do you have installed? I see in composer.lock file we have 9.1.5.

**Comment by revgovmade** [comment]
I have **PhpUnit v9.3.10**.

However, I would like to stress, even though you got a different error, this is still a problem and needs to be resolved: https://github.com/magento/magento2/issues/30146#issuecomment-697179224

The reason you have produced a different error to mine, might be because the command you ran, was slightly different to mine.

The command you ran: `../../../vendor/phpunit/phpunit/phpunit --configuration phpunit.xml ./testsuite/Magento/IntegrationTest.php`
The command I ran: `../../../vendor/phpunit/phpunit/phpunit --configuration phpunit.xml`

**Comment by ihor-sviziev** [comment]
@revgovmade sure, the issue is present, it should be resolved. I ran just single test because I saw it in your exception log that it fails.

**Comment by revgovmade** [comment]
Thanks. Getting this fixed is actually really important to us. How long do issues like this tend to get resolved? It feels like a big deal, because anyone practicing TDD can't do that with Magento 2.4 at the moment.

**Comment by ihor-sviziev** [comment]
@revgovmade as a workaround I think for now you could run only tests that you need to run

**Comment by revgovmade** [comment]
Unfortunately, there is no work around. This issue is persistent for **all** tests. I created my own test and it still has the same issue. The problem is with the Magento Integration Framework itself. From my understanding, no one can use the Magento Integration Testing Framework until the [class](https://github.com/magento/magento2/blob/2.4-develop/dev/tests/integration/testsuite/Magento/IntegrationTest.php) has been refactored to support PhpUnit 9.x.x. 

The only workaround I can see is using another Integration Testing Framework, but from my knowledge I don't know any which support Magento2.

**Comment by ihor-sviziev** [comment]
@revgovmade here is my PR that should fix this issue. You could apply it as a patch. https://github.com/magento/magento2/pull/30170

**Comment by revgovmade** [comment]
@ihor-sviziev Thanks so much for the fix. It looks like it has done the trick!

**Comment by magento-engcom-team** [comment]
Hi @revgovmade. Thank you for your report. 
The issue has been fixed in magento/magento2#30170 by @ihor-sviziev in 2.4-develop branch
Related commit(s):
 - [62cad122f33060c21b471c6589d8f4389a85a673](https://github.com/magento/magento2/commit/62cad122f33060c21b471c6589d8f4389a85a673)

The fix will be available with the upcoming 2.4.2 release.
